components/script/dom/bluetooth.rs

```diff
@@ -59,41 +61,151 @@ impl Bluetooth {
         let global_ref = global_root.r();
         global_ref.as_window().bluetooth_thread()
     }
+
+    // https://webbluetoothcg.github.io/web-bluetooth/#request-bluetooth-devices
+    fn request_bluetooth_devices(&self,
+                                 filters: &Option<Vec<BluetoothRequestDeviceFilter>>,
+                                 optional_services: &Option<Vec<BluetoothServiceUUID>>)
+                                 -> Fallible<Root<BluetoothDevice>> {
+        // TODO: Step 1: Triggered by user activation.
+
+        // Step 2.
+        let option = try!(convert_request_device_options(filters, optional_services, self.global().r()));
+
+        // TODO: Step 3-5: Impelemnt the permission API.
+
+        // Note: Steps 6-8 are implemented in
+        // components/net/bluetooth_thread.rs in request_device function.
+        let (sender, receiver) = ipc::channel().unwrap();
+        self.get_bluetooth_thread().send(BluetoothMethodMsg::RequestDevice(option, sender)).unwrap();
+        let device = receiver.recv().unwrap();
+
+        // TODO: Step 9-10: Impelemnt the permission API.
+
+        // Step 11: This step is optionable.
+
+        // Step 12-13.
+        match device {
+            Ok(device) => {
+                let ad_data = BluetoothAdvertisingData::new(self.global().r(),
+                                                            device.appearance,
+                                                            device.tx_power,
+                                                            device.rssi);
+                Ok(BluetoothDevice::new(self.global().r(),
+                                        DOMString::from(device.id),
+                                        device.name.map(DOMString::from),
+                                        &ad_data))
+            },
+            Err(error) => {
+                Err(Error::from(error))
+            },
+        }
+
+    }
 }
 
-fn canonicalize_filter(filter: &BluetoothScanFilter, global: GlobalRef) -> Fallible<BluetoothScanfilter> {
-    if filter.services.is_none() && filter.name.is_none() && filter.namePrefix.is_none() {
-        return Err(Type(FILTER_ERROR.to_owned()));
+// https://webbluetoothcg.github.io/web-bluetooth/#request-bluetooth-devices
+fn convert_request_device_options(filters: &Option<Vec<BluetoothRequestDeviceFilter>>,
+                                  optional_services: &Option<Vec<BluetoothServiceUUID>>,
+                                  global: GlobalRef)
+                                  -> Fallible<RequestDeviceoptions> {
+    // Step 2.2: There is no requiredServiceUUIDS, we scan for all devices.
+    let mut uuid_filters = vec!();
```

Comment:
can we optimize this so `uuid_filters` is created only if `!filters.is_empty()`?